### ISSUE TITLE
Lifting non-free modules

### ISSUE THREAD

**Issue by jchen419** [Macaulay2/M2#1522]
Currently lift(Module, Ring) gives an error if the module is not free:
```
i1 : R = QQ[x,y,z,w,u]/(u^4 - x*y*z*w)

o1 = R

o1 : QuotientRing

i2 : S = R/(x^3 - y^2*z)

o2 = S

o2 : QuotientRing

i3 : T = S/(y*w^2 - z*x^2)

o3 = T

o3 : QuotientRing

i4 : (N1, N2) = (coker gens ideal (T_0^4,T_1^4), comodule ideal (T_2^6,T_3^7))

o4 = (cokernel | xy2z y4 |, cokernel | z6 w7 |)

o4 : Sequence

i5 : M = N1 ++ N2

o5 = cokernel | xy2z y4 0  0  |
              | 0    0  z6 w7 |

                            2
o5 : T-module, quotient of T

i6 : lift(M, S)
stdio:6:1:(3): error: lift: expected module to be free
```
I suggest some code like the following be used to lift non-free modules:
```
i7 : nonFreeLift = method()

o7 = nonFreeLift

o7 : MethodFunction

i8 : nonFreeLift (Module, Ring) := Module => (M, S) -> (
          ringRel := presentation ring M ** S;
          liftRingRel := id_(lift(target relations M, S)) ** ringRel;
          subquotient(lift(gens M, S), lift(relations M, S) | liftRingRel)
     )

o8 = Module => -*Function[stdio:8:48-11:58]*-

o8 : Option

i9 : nonFreeLift (M, S)

o9 = subquotient (| 1 0 |, | xy2z y4 0  0  -x2z+yw2 0 0 0        0 0 |)
                  | 0 1 |  | 0    0  z6 w7 0        0 0 -x2z+yw2 0 0 |

                               2
o9 : S-module, subquotient of S
```
One may want to minimize the presentation before returning it (in the example above, the module is not minimally presented). Ideally, something like this would be the actual code that is run (cf. #1477) when `lift` is called for a non-free module.

**Comment by DanGrayson** [comment]
Can you also produce the isomorphism from o9**T to M and verify it's an isomorphism?  That isomorphism would presumably be needed in applications, unless it's an identity map.

**Comment by moorewf** [comment]
Forgive my ignorance, but isn't this covered by pushForward(phi,M) where
phi is the projection map from S to T?  I get the same answer in your
example (after pruning the result of pushForward).

**Comment by DanGrayson** [comment]
Ah, so that's what's going on -- this works because T is a quotient ring of S.

**Comment by jchen419** [comment]
@DanGrayson: Right, I'm only thinking of `lift`ing a module over a quotient ring to the original ring.

**Comment by jchen419** [comment]
@moorewf: It might be that this functionality is subsumed by `pushForward`, in which case that's great (I think `pushForward` should be more general anyways). Looking at the [documentation](https://faculty.math.illinois.edu/Macaulay2/doc/Macaulay2-1.16/share/doc/Macaulay2/Macaulay2Doc/html/_push__Forward_lp__Ring__Map_cm__Module_rp.html) though, it seems `pushForward` expects everything to be homogeneous (which the code above does not), and will give an error if this is not the case (although perhaps this can be fixed).

**Comment by DanGrayson** [comment]
In the case where T is a quotient of S, lift should just call pushForward, which apparently can be improved.

**Comment by mahrud** [comment]
Related crash:
```m2
R = QQ[x,y,z,w,u]/(u^4 - x*y*z*w)
S = R/(x^3 - y^2*z)
T = S/(y*w^2 - z*x^2)
f = map(T, S)
kernel inverse f -- internal error: incorrect Hilbert function given, aborting
```

**Comment by DanGrayson** [comment]
> Related crash:
> 
> ```m2
> R = QQ[x,y,z,w,u]/(u^4 - x*y*z*w)
> S = R/(x^3 - y^2*z)
> T = S/(y*w^2 - z*x^2)
> f = map(T, S)
> kernel inverse f -- internal error: incorrect Hilbert function given, aborting
> ```

In addition to giving the correct Hilbert function, it would be good to fix the engine code so when an incorrect Hilbert function is given at top level, it is a simple error, not a crash.  That might be worth a separate issue.

**Comment by mahrud** [comment]
@jchen419 your original example can now be computed with `pushForward(map(T, S), M)`.
